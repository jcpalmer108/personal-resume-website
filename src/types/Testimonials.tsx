import { Sections } from "./content"

export type TestimonialProps = {
  content: Sections | undefined
}

export type DotProps = {
  selected: boolean
}

export type ArrowProps = {
  desktop?: boolean | undefined
}
