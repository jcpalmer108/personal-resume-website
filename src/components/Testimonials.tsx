import { TestimonialProps } from "../types/Testimonials"
import { 
  Wrapper, 
  Quote, 
  Person, 
  Title, 
  Dots, 
  ArrowLeft, 
  ArrowRight, 
  Arrows, 
  Circle,
  Option,
  Dot,
  Content
} from "../styles/components/Testimonials"
import Section from "./Section"
import arrowLeft from "../assets/images/arrow-left-pink.svg"
import arrowRight from "../assets/images/arrow-right-pink.svg"
import { useState } from "react"

export default function Testimonials({ content }: TestimonialProps) {
  let [ selectedIndex, setSelectedIndex ] = useState(0)
  const quotes = content?.subSection?.quotes

  function updateIndex (moveForward: boolean) {
    if(moveForward) {
      return quotes && selectedIndex >= (quotes.length - 1) ? 
        setSelectedIndex(0) : 
        setSelectedIndex(selectedIndex + 1);
    } else {
      return quotes && selectedIndex <= 0 ? 
      setSelectedIndex(quotes.length - 1) : 
      setSelectedIndex(selectedIndex - 1);  
    }
  }
  
  return (
    <Section bottom>
      <Wrapper id="testimonials" data-testid="Testimonials">
        <ArrowLeft 
          onClick={() => { updateIndex(false) }} 
          src={arrowLeft} 
          alt="arrow left"
          data-testid="arrow left"
          desktop
        />
        <Content>
          <h2>Testimonials</h2>
          <Quote data-testid="Quote">{quotes && quotes[selectedIndex].quote}</Quote>
          <Person data-testid="Person">{quotes && quotes[selectedIndex].person}</Person>
          <Title data-testid="Title">{quotes && quotes[selectedIndex].title}</Title>
          <Dots data-testid="Dots">
            {quotes && quotes.map((item, index) => {
              return (<Dot selected={selectedIndex === index} key={`Dots ${index + 1} `} />)
            })}
          </Dots>
        </Content>
        <ArrowRight 
          onClick={() => { updateIndex(true) }} 
          src={arrowRight} 
          alt="arrow right" 
          data-testid="arrow right"
          desktop 
        />
        <Arrows>
          <Option onClick={() => { updateIndex(false) }} >
            <Circle />
            <ArrowLeft 
              src={arrowLeft} 
              alt="arrow left" 
              data-testid="arrow left"
            />
          </Option>
          <Option onClick={() => { updateIndex(true) }}>
            <Circle />
            <ArrowRight data-testid="arrow right" src={arrowRight} alt="arrow right"/>
          </Option>
        </Arrows>
      </Wrapper>
    </Section>
  )
  
}