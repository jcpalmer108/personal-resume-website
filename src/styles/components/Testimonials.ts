import styled from "styled-components"
import { desktopBreakpoint, off_black, redstone, silver_polish, tabletBreakpoint } from "../variables"
import { DotProps, ArrowProps } from "../../types/Testimonials"

export const Wrapper = styled.div`
  display: flex;
  flex-direction: column;
  text-align: center;
  width: 100%;

  @media (min-width: ${tabletBreakpoint}px) {
    flex-direction: row;
    justify-content: space-between;
  }
`

export const Content = styled.div`
  display: flex;
  flex-direction: column;
  text-align: center;

  @media (min-width: ${tabletBreakpoint}px) {
    max-width: 523px;
  }

  @media (min-width: ${desktopBreakpoint}px) {
    max-width: 624px;
  }
`

export const Quote = styled.p`
  margin: 30px 0px;
  font-size: 15px;
  font-family: Gilroy-Regular;
  height: fit-content; // update to change total height of section
  line-height: 26px;

  @media (min-width: ${tabletBreakpoint}px) {
    font-size: 20px;
    margin: 40px 0px;
    line-height: 34px;
  }

  @media (min-width: ${desktopBreakpoint}px) {
    font-size: 25px;
    margin: 60px 0px;
    line-height: 44px;
  }
`

export const Person = styled.div`
  font-size: 16;
  font-family: Gilroy-Medium;
  margin-bottom: 10px;

  @media (min-width: ${tabletBreakpoint}px) {
    font-size: 24px;
    margin-bottom: 14px;
  }

  @media (min-width: ${desktopBreakpoint}px) {
    font-size: 25px;
    margin-bottom: 11px;
  }
`

export const Title = styled.div`
  font-size: 12px;
  font-family: Gilroy-Regular;
  margin-bottom: 20px;

  @media (min-width: ${tabletBreakpoint}px) {
    font-size: 18px;
    margin-bottom: 30px;
  }

  @media (min-width: ${tabletBreakpoint}px) {
    font-size: 25px;
    margin-bottom: 55px;
  }
`

export const Dots = styled.div`
  display: flex;
  align-items: center;
  justify-content: center;
`

export const Dot = styled.div<DotProps>`
  height: 7px;
  width: 7px;
  margin: 0px 7.5px;
  background: ${props => props.selected ? off_black : silver_polish};
  border-radius: 50%;

  @media (min-width: ${tabletBreakpoint}px) {
    height: 10px;
    width: 10px;
    margin: 0px 9px;
  }

`


export const Arrows = styled.div`
  display: flex;
  align-items: center;
  justify-content: center;
  column-gap: 20px;
  margin-top: 30px;

  @media (min-width: ${tabletBreakpoint}px) {
    display: none; 
  }
`

export const ArrowLeft = styled.img<ArrowProps>`
  display: ${ props => props.desktop && props.desktop === true ? 'none' : 'block'}; 
  width: 19px;

  @media (min-width: ${tabletBreakpoint}px) {
    display: ${ props => props.desktop && props.desktop === true ? 'block' : 'none'}; 
    width: 40px;
  }
`

export const ArrowRight = styled.img<ArrowProps>`
  display: ${ props => props.desktop && props.desktop === true ? 'none' : 'block'}; 
  width: 19px;

  @media (min-width: ${tabletBreakpoint}px) {
    display: ${ props => props.desktop && props.desktop === true ? 'block' : 'none'}; 
    width: 40px;
  }
`

export const Circle = styled.div`
  background: ${redstone};
  opacity: 10%;
  height: 40px;
  width: 40px;
  border-radius: 50%;
  z-index: -1;
  position: absolute;
`
export const Option = styled.div`
  height: 40px;
  width: 40px;
  display: flex;
  justify-content: center;
  align-items: center;
`