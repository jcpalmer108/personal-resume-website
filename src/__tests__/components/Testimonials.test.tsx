import { fireEvent, render, screen } from '@testing-library/react'
import Testimonials from '../../components/Testimonials.tsx'

const params = {
  key: "testimonials",
  label: "Testimonials",
  title: "Test title",
  description: [
    "test description"
  ],
  inMenu: true,
  subSection: {
    quotes: [
      {
        quote: "this is a test quote",
        person: "Jane Doe",
        title: "Manager at Slalom Conuslting LLC"      
      },
      {
        quote: "this is a test quote",
        person: "John Doe",
        title: "Analyst at Slalom Conuslting LLC"      
      },
    ]
  }
}

describe('Testimonials', () => {
  test('renders if only required params are passed in', () => {
    render(<Testimonials content={params} />)

    expect(screen.getByTestId('Testimonials')).toMatchSnapshot()
    expect(screen.getByTestId('Quote')).toHaveTextContent(params.subSection.quotes[0].quote)
    expect(screen.getByTestId('Person')).toHaveTextContent(params.subSection.quotes[0].person)
    expect(screen.getByTestId('Title')).toHaveTextContent(params.subSection.quotes[0].title)
    expect(screen.getByTestId('Dots').childNodes).toHaveLength(params.subSection.quotes.length)
    expect(screen.getByRole('img', { name: 'arrow left' })).toBeTruthy()
    expect(screen.getByRole('img', { name: 'arrow right' })).toBeTruthy()
  })

  test('expect arrow left to decrement, and arrow right to increment the selected index', () => {
    render(<Testimonials content={params} />)
    expect(screen.getByTestId('Quote')).toHaveTextContent(params.subSection.quotes[0].quote)
    fireEvent.click(screen.getByRole('img', { name: 'arrow right' }))
    expect(screen.getByTestId('Quote')).toHaveTextContent(params.subSection.quotes[1].quote)
    fireEvent.click(screen.getByRole('img', { name: 'arrow left' }))
    expect(screen.getByTestId('Quote')).toHaveTextContent(params.subSection.quotes[0].quote)
  })
})