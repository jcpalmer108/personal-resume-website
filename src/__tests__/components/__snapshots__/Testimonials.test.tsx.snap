// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Testimonials renders if only required params are passed in 1`] = `
<div
  class="sc-bczRLJ gMOOza"
  data-testid="Testimonials"
>
  <img
    alt="arrow left"
    class="sc-ftvSup bIKhme"
    data-testid="arrow left"
    src="arrow-left-pink.svg"
  />
  <div
    class="sc-gsnTZi bUnqpQ"
  >
    <h2>
      Testimonials
    </h2>
    <p
      class="sc-dkzDqf fysNox"
      data-testid="Quote"
    >
      this is a test quote
    </p>
    <div
      class="sc-hKMtZM juqmlw"
      data-testid="Person"
    >
      Jane Doe
    </div>
    <div
      class="sc-eCYdqJ hEJBqo"
      data-testid="Title"
    >
      Manager at Slalom Conuslting LLC
    </div>
    <div
      class="sc-jSMfEi iOloSv"
      data-testid="Dots"
    >
      <div
        class="sc-gKXOVf ehpXSY"
      />
      <div
        class="sc-gKXOVf bjLeFz"
      />
    </div>
  </div>
  <img
    alt="arrow right"
    class="sc-papXJ QsnDr"
    data-testid="arrow right"
    src="arrow-right-pink.svg"
  />
  <div
    class="sc-iBkjds hsMbSL"
  >
    <div
      class="sc-kDDrLX hksPpZ"
    >
      <div
        class="sc-jqUVSM gjlCYy"
      />
      <img
        alt="arrow left"
        class="sc-ftvSup hdhNBy"
        data-testid="arrow left"
        src="arrow-left-pink.svg"
      />
    </div>
    <div
      class="sc-kDDrLX hksPpZ"
    >
      <div
        class="sc-jqUVSM gjlCYy"
      />
      <img
        alt="arrow right"
        class="sc-papXJ bnbZLX"
        data-testid="arrow right"
        src="arrow-right-pink.svg"
      />
    </div>
  </div>
</div>
`;
